File: unity/RNIntegration/unity/RNIntegration/Assets/Plugins/ReactNative/Source/UnityMessage.cs
using Newtonsoft.Json.Linq;
using System;
using UnityEngine.Scripting;

namespace ReactNative
{
    /// <summary>
    /// Describes unity message entity.
    /// </summary>
    [Preserve]
    public struct UnityMessage
    {
        /// <summary>
        /// The message channel ID.
        /// </summary>
        /// <remarks>
        /// The message channel ID can be used to identify target (listener) for the message.
        /// </remarks>
        public string id;

        /// <summary>
        /// Type of the message (Undefined | Request | Response).
        /// </summary>
        public int type;

        /// <summary>
        /// Optional unique identifier of the message.
        /// </summary>
        /// <remarks>
        /// This is used to route response message to the source request message.
        /// When message does not expect any response this message shall be empty.
        /// </remarks>
        public int? uuid;

        /// <summary>
        /// Optional data of the message.
        /// </summary>
        public JObject data;

        /// <summary>
        /// Gets a boolean flag indicating whether this is a simple message (no response expected).
        /// </summary>
        public bool IsSimple => !this.uuid.HasValue || this.type == (int)UnityMessageType.Default;

        /// <summary>
        /// Gets a boolean flag indicating whether this is a request message.
        /// </summary>
        public bool IsRequest => this.uuid.HasValue && this.type >= (int)UnityMessageType.Request;

        /// <summary>
        /// Gets a boolean flag indicating whether this is a response, cancel or error message.
        /// </summary>
        public bool IsRequestCompletion => this.uuid.HasValue && (this.type == (int)UnityMessageType.Response || this.type == (int)UnityMessageType.Error || this.type == (int)UnityMessageType.Canceled);

        /// <summary>
        /// Gets a boolean flag indicating whether this is a response message.
        /// </summary>
        public bool IsResponse => this.uuid.HasValue && this.type == (int)UnityMessageType.Response;

        /// <summary>
        /// Gets a boolean flag indicating whether this is an error message.
        /// </summary>
        public bool IsError => this.uuid.HasValue && this.type == (int)UnityMessageType.Error;

        /// <summary>
        /// Gets a boolean flag indicating whether this is a cancellation request.
        /// </summary>
        public bool IsCancel => this.uuid.HasValue && this.type == (int)UnityMessageType.Cancel;

        /// <summary>
        /// Gets a boolean flag indicating whether this is a cancellation notification.
        /// </summary>
        public bool IsCanceled => this.uuid.HasValue && this.type == (int)UnityMessageType.Canceled;


        /// <summary>
        /// Converts message type to a given enum type.
        /// </summary>
        /// <typeparam name="T">The type to use for conversion.</typeparam>
        /// <returns>The data in a given type.</returns>
        public T GetType<T>() where T : Enum
            => (T)Enum.ToObject(typeof(T), this.type);

        /// <summary>
        /// Converts message data to a given type.
        /// </summary>
        /// <typeparam name="T">The type to use for conversion.</typeparam>
        /// <returns>The data in a given type.</returns>
        public T GetData<T>()
            => this.data != null ? this.data.ToObject<T>() : default(T);

        /// <summary>
        /// Tries to convert message data to a given type.
        /// </summary>
        /// <typeparam name="T">The type to use for conversion.</typeparam>
        /// <param name="result">Result of the conversion</param>
        /// <returns>Boolean flag indicating whether conversion was successfull.</returns>
        public bool TryGetData<T>(out T result)
        {
            if (this.data != null)
            {
                try
                {
                    result = this.data.ToObject<T>();
                    return true;
                }
                catch (Exception e)
                {
                    UnityEngine.Debug.unityLogger.LogError("messaging", e);
                }
            }

            result = default(T);
            return false;
        }

        /// <summary>
        /// Converts message data to minimized JSON.
        /// </summary>
        /// <returns>The message in minimized JSON.</returns>
        public override string ToString()
            => this.ToString(false);

        /// <summary>
        /// Converts message data to formatted JSON.
        /// </summary>
        /// <returns>The message in formatted JSON.</returns>
        public string ToString(bool formatted)
            => ToStringInternal(
                (o) => string.Empty, // TODO formatted ? JSON.ToNiceJSON : (Func<object, string>)JSON.ToJSON,
                new
                {
                    this.id,
                    this.type,
                    this.uuid,
                    this.data
                });

        private static string ToStringInternal(Func<object, string> serializer, object data)
            => serializer(data);
    }
}

File: src/UnityMessage.ts
export enum UnityMessageType {
    Default = 0,
    Response = 1,
    Error = 2,
    Cancel = 3,
    Canceled = 4,
    Request = 9,
}

export const UnityMessagePrefix = '@UnityMessage@';

export interface IUnityMessage<TType extends number = UnityMessageType, TData = any> {
    readonly id: string;
    readonly type: TType;
    readonly data?: TData;
}

export interface UnityMessage {
    readonly id: string;
    readonly type: UnityMessageType;
    readonly uuid?: number;
    readonly data?: any;

    isSimple(): boolean;
    isRequest(): boolean;
    isRequestCompletion(): boolean;
    isResponse(): boolean;
    isCancel(): boolean;
    isError(): boolean;
}

export class UnityMessageImpl implements UnityMessage {
    private m_json: UnityMessage;

    constructor(message: string) {
        message = message.replace(UnityMessagePrefix, '');
        this.m_json = JSON.parse(message) as UnityMessage;;
    }

    public get id(): string {
        return this.m_json.id;
    }
    public get type(): UnityMessageType {
        return this.m_json.type;
    }
    public get uuid(): number | undefined {
        return this.m_json.uuid;
    }
    public get data(): any | undefined {
        return this.m_json.data;
    }

    public isSimple(): boolean {
        return this.uuid === undefined && this.type === UnityMessageType.Default;
    }

    public isRequest(): boolean {
        return this.uuid !== undefined && this.type >= UnityMessageType.Request;
    }

    public isRequestCompletion(): boolean {
        return this.uuid !== undefined && (this.type === UnityMessageType.Response || this.type === UnityMessageType.Canceled || this.type === UnityMessageType.Error);
    }

    public isResponse(): boolean {
        return this.uuid !== undefined && this.type === UnityMessageType.Response;
    }

    public isCancel(): boolean {
        return this.uuid !== undefined && this.type === UnityMessageType.Cancel;
    }

    public isCanceled(): boolean {
        return this.uuid !== undefined && this.type === UnityMessageType.Canceled;
    }

    public isError(): boolean {
        return this.uuid !== undefined && this.type === UnityMessageType.Error;
    }

    public static isUnityMessage(message: string) {
        if (message.startsWith(UnityMessagePrefix)) {
            return true;
        } else {
            return false;
        }
    }
}
